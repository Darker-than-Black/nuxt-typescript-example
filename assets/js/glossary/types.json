{
  "String": "string",
  "Array": "array",
  "Number": "number",
  "Object": "object"
}
