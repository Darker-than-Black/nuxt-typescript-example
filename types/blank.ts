export interface BlankItemField {
  title: string,
  path: string,
  itemClass?: string,
  countryPath?: string
}

export interface BlankItemList extends BlankItemField {
  items: string[]
}

export interface BlankItemConfig {
  type: string,
  path: string,
  title?: string,
  countryPath?: string,
  width?: string,
  class?: string,
  headerClass?: string,
  itemsContentClass?: string,
  itemListClass?: string,
  items: BlankItemField[],
  icons?: BlankItemList[]
}
