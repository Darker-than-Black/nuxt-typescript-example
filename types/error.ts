export type HttpError = {
  field: string,
  message: string
}

export type ErrorHandler = {
  statuses: Array<number>,
  return: any
}

export type ErrorConfig = {
  returnValue?: any,
  errorsList?: Array<ErrorHandler>
}

export type ErrorCode = {
  error?: number
}
