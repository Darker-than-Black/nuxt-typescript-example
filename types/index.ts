export * from './input'
export * from './router'
export * from './error'
export * from './table'
export * from './link'
export * from './blank'

// api
export * from './api/declaration/declarations'
export * from './api/declaration/declaration'
export * from './api/car-goods'
export * from './api/entry-journal/entry-journal'
export * from './api/entry-journal/entry-journal-ticket'
export * from './api/dictionary'

// UI
export * from './UI'

// tests
export * from './unit-test'
