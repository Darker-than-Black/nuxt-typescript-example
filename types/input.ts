export type InputValue = number|string|boolean

export type MultiSelectValue = InputValue[]

export type SelectOptions = SelectOption[]

export type SelectOption = {
  value: InputValue,
  text: string
}

export type TableRowValue = {
  [key: string]: FormFieldValue
}

export type FormFieldValue = InputValue|MultiSelectValue[]|TableRowValue[]|object

export interface IFormFieldComponentProps {
  value?: FormFieldValue,
  readonly label?: string,
  readonly disabled?: boolean,
  readonly inputClasses?: string[]|string,
  // input attrs
  readonly type?: string,
  readonly placeholder?: string,
  // select attrs
  readonly options?: SelectOptions,
  readonly filterKey?: string,
  // for validate
  readonly vid?: string,
  readonly rules?: string,
  // for data list
  readonly listLink?: string,
}

export interface IFactoryFieldConfig extends IFormFieldComponentProps {
  // ~/assets/js/glossary/field-types.json
  readonly fieldType: string,
  readonly defaultValue: FormFieldValue,
  readonly invisible?: boolean,
  readonly fields?: {
    [key: string]: IFactoryFieldConfig
  }
}

export interface IFactoryFieldsConfig {
  [key: string]: IFactoryFieldConfig
}
