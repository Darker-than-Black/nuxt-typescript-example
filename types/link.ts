export type Links = {
  [key: string]: Link
}

export type Link = {
  text: string,
  icon?: string,
  path: string,
  subLinks?: Links
}
