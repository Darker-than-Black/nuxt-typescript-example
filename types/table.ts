export type TableConfig = {
  size?: number,
  totalElements?: number,
  totalPages?: number,
  number?: number // page number
}

// Doc: https://bootstrap-vue.org/docs/components/table#field-definition-reference
export type TableColConfig = {
  key: string,
  label: string,
  sortable: boolean,
  tdClass: string[] | string,
  thClass: string[] | string
}

export type TableSortEvent = {
  apiUrl: string,
  currentPage: number,
  filter: any
  perPage: number,
  sortBy: string
  sortDesc: boolean
}
