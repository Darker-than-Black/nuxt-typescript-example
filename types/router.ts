import { InputValue, MultiSelectValue } from '~/types/input'

export type QueryParam = MultiSelectValue | InputValue

export type QueryParams = {
  [key: string]: QueryParam
}

export type QueryParamsConfig = {
  [key: string]: {
    defaultValue: QueryParam,
    type: string
  }
}
