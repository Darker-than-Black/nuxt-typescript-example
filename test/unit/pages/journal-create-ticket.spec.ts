import VueRouter from 'vue-router'
import BootstrapVue from 'bootstrap-vue'
import VueTestInstance from '~/test/factory/vue-instance'
import CreateTicketPage from '~/pages/journal/create-ticket.vue'
import EntryJournalCreateTicketModule from '~/store/entry-journal/create-ticket'
import DictionaryModule from '~/store/dictionary'
import ConvertFieldsConfigToDefaultValue from '~/assets/js/utils/convert-fields-config-to-default-value'
import CONFIG from '~/assets/js/glossary/form/create-ticker'
import LINKS from '~/assets/js/glossary/links.json'
import VehicleTypesMock from '~/test/api/vehicle-types.json'
import { CreateTicket, CreateTicketResponse, VehicleType } from '~/types'

const dataDefaultValue: CreateTicket = ConvertFieldsConfigToDefaultValue(CONFIG) as CreateTicket

jest.mock('~/api', () => ({ GET_VEHICLE_TYPES: jest.fn(() => VehicleTypesMock) }))

describe('create-ticket.vue', () => {
  let wrapper: any
  const router = new VueRouter()

  beforeEach(() => {
    wrapper = VueTestInstance(CreateTicketPage, {
      modules: [BootstrapVue],
      storeModules: {
        'entry-journal/create-ticket': EntryJournalCreateTicketModule,
        dictionary: DictionaryModule
      },
      options: {
        router,
        stubs: { ValidationObserver: { template: '<div><slot /></div>' } },
        mocks: {
          $router: { push: jest.fn() }
        }
      }
    })
  })

  it('initialize correctly', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should call getDictionaries() when created', () => {
    expect(wrapper.vm.vehicleTypes).toEqual(VehicleTypesMock)
  })

  describe('test function: onSubmit()', () => {
    it('success validation', async (done) => {
      wrapper.vm.createTicket = jest.fn()
      wrapper.vm.$refs.form.validate = jest.fn(() => {
        return new Promise((resolve: any) => {
          resolve(true)
        })
      })

      wrapper.vm.onSubmit()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.createTicket).toHaveBeenCalled()
      done()
    })

    it('fail validation', async (done) => {
      wrapper.vm.createTicket = jest.fn()
      wrapper.vm.$refs.form.validate = jest.fn(() => {
        return new Promise((resolve: any) => {
          resolve(false)
        })
      })

      wrapper.vm.onSubmit()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.createTicket).not.toHaveBeenCalled()
      done()
    })
  })

  it('check clean module, when close page', () => {
    wrapper.vm.cleanModule = jest.fn()
    wrapper.destroy()
    expect(wrapper.vm.cleanModule).toHaveBeenCalled()
  })

  it('test function: onUpdateProductsCount()', () => {
    wrapper.setData({ data: { ...wrapper.vm.data, products: [{ number: '123' }, { number: '123' }] } })
    wrapper.vm.onUpdateProductsCount()
    expect(wrapper.vm.data.products).toEqual(dataDefaultValue.products)
  })

  describe('test function: onUpdateVehicleType()', () => {
    it('unknown type name', () => {
      wrapper.vm.onUpdateVehicleType('asd')
      expect(wrapper.vm.data.vehicleCard.vehicleTypeId).toBe(dataDefaultValue.vehicleCard.vehicleTypeId)
    })

    it('known type name', () => {
      const type: VehicleType = wrapper.vm.vehicleTypes[0]
      wrapper.vm.onUpdateVehicleType(type.name)
      expect(wrapper.vm.data.vehicleCard.vehicleTypeId).toBe(type.id)
    })
  })

  describe('test function: onUpdateResponse()', () => {
    const responseSuccess: CreateTicketResponse = { cardNumber: '1234567890' }
    const responseError: CreateTicketResponse = {
      code: 400,
      title: 'Validate Error',
      violations: [
        { field: 'products', message: 'Fields is required' }
      ]
    }

    it('success created ticket', () => {
      wrapper.vm.message = jest.fn()
      wrapper.vm.onUpdateResponse(responseSuccess)
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith({ path: LINKS.journal.path })
      expect(wrapper.vm.message).toHaveBeenCalled()
    })

    it('fail created ticker', () => {
      wrapper.vm.$refs.form.setErrors = jest.fn()
      wrapper.vm.onUpdateResponse(responseError)
      expect(wrapper.vm.$refs.form.setErrors).toHaveBeenCalled()
    })

    it('another server response', () => {
      wrapper.vm.message = jest.fn()
      wrapper.vm.$refs.form.setErrors = jest.fn()
      wrapper.vm.onUpdateResponse({})
      expect(wrapper.vm.$refs.form.setErrors).not.toHaveBeenCalled()
      expect(wrapper.vm.$router.push).not.toHaveBeenCalled()
      expect(wrapper.vm.message).not.toHaveBeenCalled()
    })
  })
})
