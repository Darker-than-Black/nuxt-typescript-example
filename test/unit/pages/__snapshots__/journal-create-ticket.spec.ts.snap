// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-ticket.vue should match the snapshot 1`] = `
<b-container-stub tag="div" fluid="true">
  <b-row-stub tag="div">
    <b-col-stub cols="3" tag="div" class="mt-5">
      <b-row-stub tag="div">
        <b-col-stub cols="12" tag="div">
          <div class="icon">
            +
          </div>
        </b-col-stub>
        <b-col-stub cols="12" tag="div">
          <h1 class="title mt-3">
            Створення талону
          </h1>
        </b-col-stub>
        <b-col-stub cols="12" tag="div">
          <p class="ticket__text">
            Nec diam ut in ipsum adipiscing pulvinar consequat tempus lectus. Sit a sagittis viverra consectetur aenean hac scelerisque mauris. Sollicitudin venenatis dui pretium lacus at blandit enim rutrum.
          </p>
        </b-col-stub>
        <b-col-stub cols="12" tag="div">
          <div class="notification success">
            <div class="notification__content">
              Позначені* поля обов’язкові для заповнення
            </div>
          </div>
        </b-col-stub>
      </b-row-stub>
    </b-col-stub>
    <b-col-stub cols="7" tag="div">
      <div tag="div" class="ticket">
        <form class="container-fluid p-0">
          <b-row-stub tag="div">
            <b-col-stub cols="12" tag="div">
              <management-table-stub value="[object Object],[object Object]" label="" placeholder="" type="text" inputclasses="" vid="" rules="" listlink="" config="[object Object]" btntext="Додати документ"></management-table-stub>
            </b-col-stub>
          </b-row-stub>
          <b-row-stub tag="div" class="mt-4">
            <b-col-stub cols="11" tag="div" class="d-flex align-items-center justify-content-between">
              <checkbox-stub value="true" label="Всі товари з документу" placeholder="" type="text" inputclasses="" vid="" rules="" listlink=""></checkbox-stub> <button type="button" class="btn btn-transparent white-space-nowrap">
                Вказати номер товару <b>+</b></button>
            </b-col-stub>
            <b-col-stub cols="12" tag="div">
              <management-table-stub value="[object Object],[object Object]" label="" placeholder="" type="text" inputclasses="" vid="" rules="" listlink="" config="[object Object]" btntext="Додати документ"></management-table-stub>
            </b-col-stub>
          </b-row-stub>
          <b-row-stub tag="div" class="mt-4">
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
          </b-row-stub>
          <b-row-stub tag="div" class="mt-4">
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="0" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <!---->
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
          </b-row-stub>
          <b-row-stub tag="div" class="mt-4">
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
            <b-col-stub cols="11" tag="div">
              <factory-field-stub value="" config="[object Object]"></factory-field-stub>
            </b-col-stub>
          </b-row-stub>
          <b-row-stub tag="div" class="mt-4">
            <b-col-stub cols="11" tag="div" class="text-right"><a class="btn btn-gray mr-2">
                Скасувати
              </a> <button type="submit" class="btn btn-blue">
                Створити талон
              </button></b-col-stub>
          </b-row-stub>
        </form>
      </div>
    </b-col-stub>
  </b-row-stub>
</b-container-stub>
`;
